fili-core/src/main/java/com/yahoo/bard/webservice/data/metric/protocol/MetricTransformer.java

```diff
@@ -0,0 +1,41 @@
+// Copyright 2020 Oath Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.data.metric.protocol;
+
+import com.yahoo.bard.webservice.data.metric.LogicalMetric;
+
+import java.util.Map;
+
+/**
+ * An interface for transforming metrics into other metrics.
+ */
+@FunctionalInterface
+public interface MetricTransformer {
+
+    /**
+     * Transform a metric using a Protocol, and signal data.
+     *
+     * @param logicalMetric  The metric to transform.
+     * @param protocol The protocol being transformed
+     * @param parameterValues  The data associated with that signal.
+     *
+     * @return A new metric based on the signal;
+     * @throws UnknownProtocolValueException if this transformer doesn't know how to accept this signal
+     */
+    LogicalMetric apply(LogicalMetric logicalMetric, Protocol protocol, Map<String, String> parameterValues)
```

Comment:
Taking Protocol here is a little code smelly to me. Protocol is tied to a MetricTransformer instance, and it seems unlikely to me that a single MetricTransformer would be able to handle multiple different Protocol contracts. Thus, consuming the applying Protocol (which should just be the Protocol that references this transformer) feels weird. See my comment on TimeAverageMetricTransformer for more in depth thoughts on MetricTransformer.

Reply:
After discussing with @michael-mclawhorn we will change Transformers to NOT close over the coreParameter, nor the contractName. The passed in Protocol will provide the core parameter to the transformer, which it can then use to pull the core parameter value out of the `parameterValues` map. Note that the Transformer WILL need to close over any auxiliary parameter names.